### PR TITLE
[Feature] Add Rayleigh phase function

### PR DESCRIPTION
**Pull request by leroyvn**
## Description

This PR adds a phase function plugin for the Rayleigh volume scattering model. The implementation is taken from the Mitsuba 1 plugin and does not account for polarisation.

*Note:* This PR branches off `master`; I can rebase it on `next` if appropriate.

## Testing

Added a test file including:

- a plugin instantiation test;
- a chi2 test.

## Checklist:

- [x] My code follows the [style guidelines](https://mitsuba2.readthedocs.io/en/latest/src/developer_guide/intro.html#introduction) of this project
- [x] My changes generate no new warnings
- [ ] My code also compiles for `gpu_*` and `packet_*` variants. If you can't test this, please leave below
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I cleaned the commit history and removed any "Merge" commits
- [x] I give permission that the Mitsuba 2 project may redistribute my contributions under the terms of its [license](https://github.com/mitsuba-renderer/mitsuba2/blob/master/LICENSE)

*Note:* Tested with the `llvm_rgb` variant (not with GPU variants).

### CODE REVIEW

**Comment by Speierers** [comment]
Hi @leroyvn,

Thanks for this PR! It looks good to me.

@dvicini is the `PhaseFunction` API changing at all in the `next` branch?

Otherwise, please go ahead an rebase this onto `next` as it won't be necessary to merge this into `master`.

Would it be possible to add an rendered image of a commun usecase of this phase function to the documentation?

**Comment by leroyvn** [comment]
Thanks, I'll rebase and update this PR. As for a rendered image, producing something good-looking is a bit of a challenge at the moment (an image of a planet with a blue atmospheric haze would be best, but I think I'm missing a spherical volume data source to do this somewhat efficiently).

File: src/phase/rayleigh.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            Log(Warn,
+                "Polarized version of Rayleigh phase function not implemented, "
+                "falling back to scalar version");
+
+        m_flags = +PhaseFunctionFlags::Anisotropic;
+    }
+
+    MTS_INLINE Float eval_rayleigh(Float cos_theta) const {
+        return (3.f / 16.f) * ek::InvPi<Float> *
+               (1.f + enoki::sqr(cos_theta));
```
Comment (Speierers) [inline src/phase/rayleigh.cpp:43]: It would be great if you could replace all instances of `enoki::` by `ek::` 😄

File: src/phase/rayleigh.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    std::pair<Vector3f, Float> sample(const PhaseFunctionContext & /* ctx */,
+                                      const MediumInteraction3f &mi,
+                                      const Point2f &sample,
+                                      Mask active) const override {
+        MTS_MASKED_FUNCTION(ProfilerPhase::PhaseFunctionSample, active);
+
+        Float z         = 2.f * (2.f * sample.x() - 1.f);
+        Float tmp       = enoki::sqrt(enoki::sqr(z) + 1.f);
+        Float A         = enoki::cbrt(z + tmp);
+        Float B         = enoki::cbrt(z - tmp);
```
Comment (Speierers) [inline src/phase/rayleigh.cpp:55]: Style-wise, I would rather not have those extra spaces before the `=`. Maybe something like this:
```diff
-     Float z         = 2.f * (2.f * sample.x() - 1.f);
-     Float tmp       = enoki::sqrt(enoki::sqr(z) + 1.f);
-     Float A         = enoki::cbrt(z + tmp);
-     Float B         = enoki::cbrt(z - tmp);
+    Float z   = 2.f * (2.f * sample.x() - 1.f);
+    Float tmp = enoki::sqrt(enoki::sqr(z) + 1.f);
+    Float A   = enoki::cbrt(z + tmp);
+    Float B   = enoki::cbrt(z - tmp);
```

File: src/phase/rayleigh.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        Float cos_theta = A + B;
+        Float sin_theta = enoki::safe_sqrt(1.0f - enoki::sqr(cos_theta));
+        auto [sin_phi, cos_phi] = enoki::sincos(ek::TwoPi<Float> * sample.y());
+
+        auto wo = Vector3f{ sin_theta * cos_phi, sin_theta * sin_phi, cos_theta };
+
+        wo        = mi.to_world(wo);
```
Comment (Speierers) [inline src/phase/rayleigh.cpp:62]: ditto
```
wo = mi.to_world(wo);
```

**Comment by leroyvn** [comment]
Hi @Speierers, I pushed style updates.

**Comment by Speierers** [comment]
Thanks @leroyvn ,

Sorry I forgot to mention that it would be great to add empty lines at the end of the plugin cpp file and the python test file 😉

**Comment by leroyvn** [comment]
Sure! Sorry I missed that, I usually add those as well!

**Comment by Speierers** [comment]
Thanks a lot @leroyvn ! 🚀

**Comment by merlinND** [comment]
@leroyvn If you're fine with having those "empty line at end of file" everywhere (I think they're quite standard), there's usually an option to do it automatically in text editors / IDEs.

**Pull request merged**